### ISSUE TITLE
Recommend the Decoupled Contrastive Learning loss

### ISSUE THREAD

**Issue by huhk-sysu** [UKPLab/sentence-transformers#1388]
Recently [Yeh at el.](https://arxiv.org/pdf/2110.06848.pdf) proposed decoupled the contrastive learning (DCL) objective function. It is an improvement over the InfoNCE (IMO, it's mostly the same as the `MultipleNegativesRankingLoss` in your library) loss. 

The DCL loss is simple, it just removes the positive pair from the denominator of InfoNCE. However, it behaves quite well, especially on small batch size.

![image](https://user-images.githubusercontent.com/14769033/150307594-32fccb6a-1e07-4cd4-921a-afc1256a795d.png)

![image](https://user-images.githubusercontent.com/14769033/150308243-506a1c15-32da-40d9-9970-7b5f45b53aaa.png)

And [here](https://github.com/raminnakhli/Decoupled-Contrastive-Learning) is an implementation.

Sadly, I haven't evaluated its effectiveness on NLP tasks. But I believe it would be helpful, since NLP models always have a smaller batch size compared to CV models.

**Comment by ScottishFold007** [comment]
I test it right now,It's okay:
![image](https://user-images.githubusercontent.com/36957508/150560631-f748bb2e-1d3e-4464-aff0-7b6f4ce9ed36.png)

**Comment by huhk-sysu** [comment]
@ScottishFold007 

Hi, have you ever compared it to other loss functions? I've done experiments on the STS dataset, but the result is not competitive. I'm not sure what my mistake is.

I have unified the losses and here's my implementation (sorry for image only since I can't copy the code):
(I assume there's always three texts in an example: the anchor, the entailment one and the contradictive one.)
![image](https://user-images.githubusercontent.com/14769033/150896595-04310591-a4a6-4631-8aac-727fb0bc76d7.png)

The three parts of `neg_mask` separately indicate that how we optimize the similarity between (anchor & anchor, anchor & positive one, anchor & negative one). When a position of `neg_mask` have a value of zero, it means we want to decrease the similarity between the associated two texts. By tuning `neg_mask` we can obtain distinct functions:

- When `neg_mask` becomes: `torch.cat([torch.ones, torch.zeros, torch.zeros])`, the loss becomes the Multiple Negatives Ranking Loss.
- When `neg_mask` becomes: `torch.cat([torch.eye, torch.zeros, torch.zeros])`, the loss becomes the normal InfoNCE Loss (with hard negative) in SimCLR.
- When `neg_mask` becomes: `torch.cat([torch.eye, torch.eye, torch.zeros])`, the loss becomes Decoupled Contrastive Learning Loss (with hard negative).

I done my experiments based on [training_nli_v2.py](examples/training/nli/training_nli_v2.py) in this repo, and I found the InfoNCE loss has the same result (0.84+) as Multiple Negatives Ranking Loss, and both are superior to the Decoupled Contrastive Learning Loss (0.80+).

**Comment by ScottishFold007** [comment]
> @ScottishFold007
> 
> Hi, have you ever compared it to other loss functions? I've done experiments on the NIL dataset, but the result is not competitive. I'm not sure what my mistake is.
> 
> I have unified the losses and here's my implementation (sorry for image only since I can't copy the code): (I assume there's always three texts in an example: the anchor, the entailment one and the contradictive one.) ![image](https://user-images.githubusercontent.com/14769033/150896595-04310591-a4a6-4631-8aac-727fb0bc76d7.png)
> 
> The three parts of `neg_mask` separately indicate that how we optimize the similarity between (anchor & anchor, anchor & positive one, anchor & negative one). When a position of `neg_mask` have a value of zero, it means we want to decrease the similarity between the associated two texts. By tuning `neg_mask` we can obtain distinct functions:
> 
> * When `neg_mask` becomes: `torch.cat([torch.ones, torch.zeros, torch.zeros])`, the loss becomes the Multiple Negatives Ranking Loss.
> * When `neg_mask` becomes: `torch.cat([torch.eye, torch.zeros, torch.zeros])`, the loss becomes the normal InfoNCE Loss (with hard negative) in SimCLR.
> * When `neg_mask` becomes: `torch.cat([torch.eye, torch.eye, torch.zeros])`, the loss becomes Decoupled Contrastive Learning Loss (with hard negative).
> 
> I done my experiments based on [training_nli_v2.py](examples/training/nli/training_nli_v2.py) in this repo, and I found the InfoNCE loss has the same result (0.84+) as Multiple Negatives Ranking Loss, and both are superior to the Decoupled Contrastive Learning Loss (0.80+).

How aboutWeightedDecoupledContrastiveLoss？ Do U use it?

**Comment by huhk-sysu** [comment]
> 

I didn't. I can have a try, but I ‘m afraid I have misunderstood the formula or implement it in a wrong way.

**Comment by huhk-sysu** [comment]
> > @ScottishFold007
> > Hi, have you ever compared it to other loss functions? I've done experiments on the NIL dataset, but the result is not competitive. I'm not sure what my mistake is.
> > I have unified the losses and here's my implementation (sorry for image only since I can't copy the code): (I assume there's always three texts in an example: the anchor, the entailment one and the contradictive one.) ![image](https://user-images.githubusercontent.com/14769033/150896595-04310591-a4a6-4631-8aac-727fb0bc76d7.png)
> > The three parts of `neg_mask` separately indicate that how we optimize the similarity between (anchor & anchor, anchor & positive one, anchor & negative one). When a position of `neg_mask` have a value of zero, it means we want to decrease the similarity between the associated two texts. By tuning `neg_mask` we can obtain distinct functions:
> > 
> > * When `neg_mask` becomes: `torch.cat([torch.ones, torch.zeros, torch.zeros])`, the loss becomes the Multiple Negatives Ranking Loss.
> > * When `neg_mask` becomes: `torch.cat([torch.eye, torch.zeros, torch.zeros])`, the loss becomes the normal InfoNCE Loss (with hard negative) in SimCLR.
> > * When `neg_mask` becomes: `torch.cat([torch.eye, torch.eye, torch.zeros])`, the loss becomes Decoupled Contrastive Learning Loss (with hard negative).
> > 
> > I done my experiments based on [training_nli_v2.py](examples/training/nli/training_nli_v2.py) in this repo, and I found the InfoNCE loss has the same result (0.84+) as Multiple Negatives Ranking Loss, and both are superior to the Decoupled Contrastive Learning Loss (0.80+).
> 
> How aboutWeightedDecoupledContrastiveLoss？ Do U use it? Notice： **weight_fn = lambda embeddings_a, embeddings_b: 2 - embeddings_a.size(0) * torch.nn.functional.softmax((embeddings_a * embeddings_b).sum(dim=1) / self.sigma, dim=0).squeeze()**

I didn't. I ‘m afraid I have misunderstood the formula of DecoupledContrastiveLoss or implement it in a wrong way, so I didn't go further. I can have a try if you want.
